src/main/java/org/opensearch/neuralsearch/query/HybridQuery.java

```diff
@@ -0,0 +1,209 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.neuralsearch.query;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Objects;
+
+import org.apache.lucene.index.IndexReader;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.search.BooleanClause;
+import org.apache.lucene.search.BooleanQuery;
+import org.apache.lucene.search.Explanation;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.MatchNoDocsQuery;
+import org.apache.lucene.search.Matches;
+import org.apache.lucene.search.MatchesUtils;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.QueryVisitor;
+import org.apache.lucene.search.ScoreMode;
+import org.apache.lucene.search.Scorer;
+import org.apache.lucene.search.Weight;
+
+/**
+ * Implementation fo Query interface for type "hybrid". It allows execution of multiple sub-queries and collect individual
+ * scores for each sub-query.
+ */
+public class HybridQuery extends Query implements Iterable<Query> {
+
+    private final List<Query> subQueries = new ArrayList<>();
+
+    public HybridQuery(Collection<Query> subQueries) {
+        Objects.requireNonNull(subQueries, "Collection of Queries must not be null");
+        this.subQueries.addAll(subQueries);
+    }
+
+    @Override
+    public Iterator<Query> iterator() {
+        return getSubQueries().iterator();
+    }
+
+    @Override
+    public String toString(String field) {
+        StringBuilder buffer = new StringBuilder();
+        buffer.append("(");
+        Iterator<Query> it = subQueries.iterator();
+        for (int i = 0; it.hasNext(); i++) {
+            Query subquery = it.next();
+            if (subquery instanceof BooleanQuery) { // wrap sub-bools in parens
+                buffer.append("(");
+                buffer.append(subquery.toString(field));
+                buffer.append(")");
+            } else buffer.append(subquery.toString(field));
+            if (i != subQueries.size() - 1) buffer.append(" | ");
+        }
+        buffer.append(")");
+        return buffer.toString();
+    }
+
+    @Override
+    public Query rewrite(IndexReader reader) throws IOException {
+        if (subQueries.isEmpty()) {
+            return new MatchNoDocsQuery("empty HybridQuery");
+        }
+
+        if (subQueries.size() == 1) {
+            return subQueries.iterator().next();
+        }
+
+        boolean actuallyRewritten = false;
+        List<Query> rewrittenSubQueries = new ArrayList<>();
+        for (Query subQuery : subQueries) {
+            Query rewrittenSub = subQuery.rewrite(reader);
+            actuallyRewritten |= rewrittenSub != subQuery;
```

Comment:
add a comment what line is doing and why we need this?

Reply:
ack